structures/lexicon.py

```diff
@@ -11,10 +11,13 @@ class Lexicon(object):
     """
     Definition 3: A lexicon is a finite set of lexical items.
     """
-    def __init__(self):
+    def __init__(self,file_name=None):
```

Comment:
Si querés que por defecto `file_name` sea igual a "lexicon.xml", podés poner directamente:
```suggestion
    def __init__(self,file_name="lexicon.xml"):
```

De esta forma, si el usuario indica otro file_name, se va a tomar el del usuario. Y, si no, se toma "lexicon.xml". Con esto podés borrar las líneas 17-20.

Reply:
Hola Macarena. Entiendo que sería más prolijo eliminar la variable `filename`, y que hay un `global filename` innecesario. Lo que no me queda claro es: al hacer este cambio, si el usuario no instancia directamente el Lexicon, ¿cómo puede saber la función `manual_derivation`, sin una variable que contenga el nombre del archivo, si pasarle o no un argumento a `Lexicon()`? El código actual también usa esa variable para mostrar en pantalla el nombre del archivo en uso.
Sí se me ocurre que podría conservar esta variable, sólo que con valor None. Y luego un `if isinstance(filename, type(None)):` etc.
Gracias!

Reply:
Hola, Facundo. Estoy un poco desorientada. ¿Puede ser que esta respuesta sea para el comentario donde te sugiero borrar las variables con `global`? La función `manual_derivation` está definida en el archivo `main.py`. Ahí, en la línea 5 ya estás definiendo la variable `filename` y asinándole un valor (_lexicon.xml_). En la línea 26 de ese mismo archivo se instancia la clase `Lexicon` con ese filename. No sé si con esto respondo a tu comentario o iba por otro lado. Avisame sin problemas y lo podemos seguir conversando.

Reply:
Mi comentario era una duda sobre la "plausibilidad" de modificar el constructor de la clase `Lexicon` y eliminar la variable `filename` de `main.py` (pero tal vez entendí mal la propuesta). Y decía, en ese caso, ¿cómo va a saber la función `manual_derivation` si pasarle un parámetro o no a la clase (y que tome el nombre del archivo por defecto)? 

En cuanto a las dos líneas `global filename`, veo bien que la que se encuentra en la función `print_grammars()` no tiene  mucho sentido. Pero en la función `main()` la agregué con la idea de poder modificar el valor de la variable desde dentro de la función (siguiendo el último ejemplo de [esta explicación](https://www.w3schools.com/python/python_variables_global.asp)). Y a su vez está definida de manera global (en la línea 5) para que su valor pueda ser accesible para `print_grammars()`, aunque acá la redefinición como global sea innecesaria.

Comprendido lo de las dos clases `Derivation` que comentabas más abajo.

Gracias!

Reply:
Hola, Facundo. Claro. Creo que algunas cosas me quedan más claras ahora.

Respecto de `main.py`, es cierto que si no lo hacés con `global`, no podés modificar la variable que está definida afuera. No me había dado cuenta. Disculpame. Una alternativa para todo eso podría ser cambiar la línea 8 por `filename="lexicon.xml"` (dentro de la función `main()`) y que la función `print_grammars()` tome ese valor. Te sugiero eso porque eso te permitiría, si quisieras, definir la función `print_grammars()` en otro archivo si quisieras (para que no te quede un `main.py` con tanta lógica) y, por otro lado, evitar el extraño diseño al tener una función que toma una variable definida externamente mientras que otras funciones necesitan tomar ese valor como un parámetro. En caso de que prefirieses seguir con el `global`, por ahí consideraría moverlo al flujo dentro del `elif` para el valor 3, dado que solo ahí la función `main()` cambia su valor. Antes no. Esto último, asumiento que dejás la variable definida también en la línea 5 tal como esta.

En cuando a la definición de la clase `Lexicon`, no me termina de quedar clara tu duda. La sugerencia que te hago es para evitar el flujo de si `file_name` es None. Cuando se quiere definir una variable con un valor por defecto, se suele hacer como te sugiero. La función `manual_derivation()` debería pasarle el valor del `filename` de todos modos porque, si el usuario elige uno que no es el que se indica por defecto, está bien pasarlo. Y si es el por defecto, no está mal tampoco. No termino de entender la lógica externa con el `None` que planteas (dentro de `manual_derivation()` si no entiendo mal). 

Si querés, podemos arreglar para hablarlo por videollamada dos minutos porque quizá es que no estoy atando todos los cabos correctamente.

Reply:
Hola Macarena! Perfecto, parece mucho más sensato definir la variable `filename` dentro de `main()` y pasársela a `print_grammars()`. Así además `print_grammars()` queda como una función más flexible (se le podrían sumar otros parámetros).

También creo que entendí lo de `file_name`. Eso me permite ahorrarme el bloque:
```
if file_name == None:
            file_name = "lexicon.xml"
        else:
            file_name = file_name
```
Que la verdad no tenía mucho sentido porque los valores por default se definen mejor como vos decís.

¿Te referías a esto? Si no, hacemos la videollamada y lo terminamos de resolver.

Gracias!

Reply:
Hola, Facundo. Disculpame la demora en la repsuesta. La notificación quedó sepultada entre otros correos y no la había visto. Efectiamente e refería a esa porción de código. Al indicarla por defecto y/o pasarla siempre desde la función que llama a la clase, te evitás el condicional.